calico_node/allocateipip/allocate_ipip_addr.go

```diff
@@ -105,40 +130,70 @@ func assignHostTunnelAddr(c *client.Client, nodename string, ipipCidrs []net.IPN
 		IPv4Pools: ipipCidrs,
 	}
 
-	if ipv4Addrs, _, err := c.IPAM().AutoAssign(args); err != nil {
+	ipv4Addrs, _, err := c.IPAM().AutoAssign(ctx, args)
+	if err != nil {
 		log.WithError(err).Fatal("Unable to autoassign an address for IPIP")
-	} else if len(ipv4Addrs) == 0 {
+	}
+
+	if len(ipv4Addrs) == 0 {
 		log.Fatal("Unable to autoassign an address for IPIP - pools are likely exhausted.")
-	} else if err = c.Config().SetNodeIPIPTunnelAddress(nodename, &ipv4Addrs[0]); err != nil {
-		log.WithError(err).WithField("IP", ipv4Addrs[0].String()).Fatal("Unable to set IPIP tunnel address")
-	} else {
-		log.WithField("IP", ipv4Addrs[0].String()).Info("Set IPIP tunnel address")
 	}
+
+	node, err := c.Nodes().Get(ctx, nodename, options.GetOptions{})
+	if err != nil {
+		log.WithError(err).Fatalf("Unable to retrieve IPIP tunnel address for cleanup. Error getting node '%s'", nodename)
+	}
+
+	node.Spec.BGP.IPv4IPIPTunnelAddr = ipv4Addrs[0].String()
+
+	// If the update fails with ResourceConflict error then retry 5 times with 1 second delay before failing.
+	for i := 0; i < 5; i++ {
+		_, err = c.Nodes().Update(ctx, node, options.SetOptions{})
+		if _, ok := err.(cerrors.ErrorResourceUpdateConflict); ok {
+			// Wait for a second and try again if there was a conflict during the resource update.
+			log.Infof("Error updating node %s: %s. Retrying.", node.Name, err)
+			time.Sleep(1 * time.Second)
+			continue
+		} else if err != nil {
+			// We hit an error, so release the IP address before exiting.
+			_, err := c.IPAM().ReleaseIPs(ctx, []net.IP{ipv4Addrs[0]})
+			if err != nil {
+				log.WithError(err).WithField("IP", ipv4Addrs[0].String()).Errorf("Error releasing IP address on faiure")
+			}
+
+			log.WithError(err).WithField("IP", ipv4Addrs[0].String()).Fatal("Unable to set IPIP tunnel address")
+		}
+		break
+	}
+
+	log.WithField("IP", ipv4Addrs[0].String()).Info("Set IPIP tunnel address")
```

Comment:
Not necessarily true - if you hit an error in the loop then you didn't set the tunnel address.

I think you would be better off doing the error check outside this retry loop.   So basically if you hit an error then release the IP address, otherwise log success.

Reply:
if you hit an error (other than updateConflict) we're logging it and exiting with `Fatal()` in line 164